lib/urn.rb

```diff
@@ -1,26 +1,88 @@
+def URN(urn)
+  URN.new(urn)
+end
+
 class URN
+  InvalidURNError = Class.new(StandardError)
+
   PATTERN = %{(?i:urn:(?!urn:)[a-z0-9][a-z0-9\-]{1,31}:} +
-            %{(?:[a-z0-9()+,-.:=@;$_!*']|%[0-9a-f]{2})+)}.freeze
+            %{(?:[a-z0-9()+,-.:=@;$_!*']|%(?:2[1-9a-f]|[3-9a-f][0-9a-f]))+)}.freeze
   REGEX = /\A#{PATTERN}\z/
 
   attr_reader :urn
   private :urn
 
+  def self.extract(str)
+    if block_given?
+      str.scan(/#{PATTERN}/) { yield $& }
+      nil
+    else
+      result = []
+      str.scan(/#{PATTERN}/) { result.push $& }
+      result
+    end
```

Comment:
I'm wondering if we should use the principle of [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) here: this seems complicated for a use case that does not yet exist so maybe we should only introduce it if we use capturing groups?